src/analysis/processing/qgsalgorithmfieldcalculator.cpp

```diff
@@ -0,0 +1,222 @@
+  std::unique_ptr< QgsProcessingParameterExpression > expression = qgis::make_unique< QgsProcessingParameterExpression> ( QStringLiteral( "FORMULA" ), QObject::tr( "Formula" ), QVariant(), QStringLiteral( "INPUT" ), false, true );
+
+  addParameter( field.release() );
+  addParameter( expression.release() );
+}
+
+QgsFields QgsFieldCalculatorAlgorithm::outputFields( const QgsFields & ) const
+{
+  return mFields;
+}
+
+QString QgsFieldCalculatorAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm computes a new vector layer with the same features of the input layer, "
+                      "but with an additional attribute. The values of this new attribute are computed from each "
+                      "feature using a mathematical formula, based on the properties and attributes of the feature." );
+}
+
+QgsFieldCalculatorAlgorithm *QgsFieldCalculatorAlgorithm::createInstance() const
+{
+  return new QgsFieldCalculatorAlgorithm();
+}
+
+
+bool QgsFieldCalculatorAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  QMap<QString, QVariant::Type> fieldTypes(
+  {
+    {"Real", QVariant::Double},
+    {"Integer", QVariant::Int},
+    {"String", QVariant::String},
+    {"Date", QVariant::Date},
+  } );
+
+  // prepare fields
+  QVariantMap fieldDefinitionMap = parameters.value( QStringLiteral( "FIELD" ) ).toMap();
+
+  if ( fieldDefinitionMap.isEmpty() )
+  {
+    fieldDefinitionMap.insert( QStringLiteral( "name" ), parameters.value( QStringLiteral( "FIELD_NAME" ) ).toString() );
+    fieldDefinitionMap.insert( QStringLiteral( "type" ), parameters.value( QStringLiteral( "FIELD_TYPE" ) ).toString() );
+    fieldDefinitionMap.insert( QStringLiteral( "length" ), parameters.value( QStringLiteral( "FIELD_LENGTH" ) ).toInt() );
+    fieldDefinitionMap.insert( QStringLiteral( "precision" ), parameters.value( QStringLiteral( "FIELD_PRECISION" ) ).toInt() );
+  }
+
+  const QString fieldType = fieldDefinitionMap.value( QStringLiteral( "type" ) ).toString();
+  const QgsField field(
+    fieldDefinitionMap.value( QStringLiteral( "name" ) ).toString(),
+    fieldTypes.value( fieldType ),
+    fieldType,
+    fieldDefinitionMap.value( QStringLiteral( "length" ) ).toInt(),
+    fieldDefinitionMap.value( QStringLiteral( "precision" ) ).toInt()
+  );
+
+  mFields = source->fields();
+  int fieldIdx = mFields.lookupField( field.name() );
+
+  if ( fieldIdx < 0 )
+    mFields.append( field );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, mFields, QgsWkbTypes::multiType( source->wkbType() ), source->sourceCrs() ) );
```

Comment:
The sink should be creatd in processAlgorithm, not prepare

Reply:
I don't need the sink at all, it is a leftover before I moved the alg to be a `QgsProcessingFeatureBasedAlgorithm`